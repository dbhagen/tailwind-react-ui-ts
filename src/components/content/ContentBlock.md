Extended by content section types
