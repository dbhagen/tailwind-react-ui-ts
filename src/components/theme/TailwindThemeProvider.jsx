import React from 'react'
import PropTypes from 'prop-types'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

const TailwindThemeProvider = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

TailwindThemeProvider.propTypes = {
  theme: PropTypes.shape({}),
  children: PropTypes.node,
}

TailwindThemeProvider.defaultProps = {
  theme: {},
  children: undefined,
}

export default TailwindThemeProvider
