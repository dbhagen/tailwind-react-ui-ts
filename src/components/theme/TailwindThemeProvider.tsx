import React from 'react'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

export type TailwindThemeProviderProps = {
  theme: {}
  children: React.ReactNode
}

const TailwindThemeProvider: React.FunctionComponent<TailwindThemeProviderProps> = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

export default TailwindThemeProvider
