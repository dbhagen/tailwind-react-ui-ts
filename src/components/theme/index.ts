export { default as defaultTheme } from './defaultTheme'
export { default as TailwindTheme } from './TailwindTheme'
export { default as TailwindThemeProvider } from './TailwindThemeProvider'
export { default as withTheme } from './withTheme'
export { default as useThemeValue } from './useThemeValue'
