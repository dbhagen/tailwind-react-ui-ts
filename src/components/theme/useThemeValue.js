import { tailwindPropToClassName } from '../tailwind'

export default (prefix, value, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
