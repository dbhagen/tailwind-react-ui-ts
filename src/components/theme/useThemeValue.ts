import { tailwindPropToClassName } from '../tailwind'

export default (prefix: string, value: string, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
