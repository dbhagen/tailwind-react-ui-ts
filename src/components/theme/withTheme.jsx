import React from 'react'

import TailwindTheme from './TailwindTheme'

const withTheme = Component => {
  const WithTheme = props => (
    <TailwindTheme.Consumer>
      {theme => <Component {...props} theme={theme} />}
    </TailwindTheme.Consumer>
  )

  WithTheme.displayName = `WithTheme(${Component.displayName})`

  return WithTheme
}

export default withTheme
