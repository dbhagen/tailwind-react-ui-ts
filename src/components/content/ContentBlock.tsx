import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { Box } from '../primitives'
import { withTheme } from '../theme'
import { getUniqueID } from '../utils'

import ContentTitle from './ContentTitle'

class ContentBlock extends PureComponent {
  constructor(props) {
    super(props)
    this.id = getUniqueID('content')
  }

  render() {
    const { theme, is, children, ...rest } = this.props

    return (
      <Box is={is} p={theme.spacing.md} aria-labelledby={this.id} {...rest}>
        {React.Children.map(children, (child, index) => {
          if (child.type === ContentTitle) {
            return React.cloneElement(child, { content: { id: this.id } })
          }

          if (index === React.Children.count(children) - 1) {
            return React.cloneElement(child, { m: { b: 0 } })
          }

          return child
        })}
      </Box>
    )
  }
}

ContentBlock.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
}

ContentBlock.defaultProps = {
  is: 'section',
  children: undefined,
}

export { ContentBlock as component }
export default withTheme(ContentBlock)
