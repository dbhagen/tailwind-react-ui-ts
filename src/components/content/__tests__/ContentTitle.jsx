import React from 'react'
import { shallow } from 'enzyme'

import { component as ContentTitle } from '../ContentBlock'
import { defaultTheme } from '../../theme'

const setup = (testProps = {}) => {
  const props = Object.assign({ theme: defaultTheme }, testProps)

  const wrapper = shallow(<ContentTitle {...props}>Title</ContentTitle>)

  return {
    props,
    wrapper,
  }
}

describe('ContentBlock', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup()

    expect(wrapper).toMatchSnapshot()
  })
})
