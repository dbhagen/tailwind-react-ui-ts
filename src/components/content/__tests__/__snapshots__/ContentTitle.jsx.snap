// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentBlock renders matching snapshot 1`] = `
<Box
  aria-labelledby="content-1"
  inline={false}
  inlineBlock={false}
  is="section"
  p={4}
>
  <Component
    key=".0"
    m={
      Object {
        "b": 0,
      }
    }
  />
</Box>
`;
